### PR TITLE
Add an Approximate funsor for use in adjoint algorithms

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #446 
pair coded with @eb8680 

This was initially called `Importance(arg, proposal, vars)` in the [design doc](https://docs.google.com/document/d/1jbpGwAJLAfTAi6IZyPUSArgXt_Km49573UI7izsDvsE/edit#).

## Tested
- [x] eager tests
- [x] MonteCarlo tests
- [x] moment_matching tests

### CODE REVIEW

File: funsor/terms.py [hunk 1]
```diff
@@ -1249,6 +1274,28 @@ def moment_matching_reduce(op, arg, reduced_vars):
     return instrument.debug_logged(arg.moment_matching_reduce)(op, reduced_vars)
 
 
+class Approximate(Funsor):
+    """
+    Function approximation wrt a semiring.
+    """
+
+    def __init__(self, sum_op, prod_op, model, guide, approx_vars):
```
Comment (fritzo) [inline funsor/terms.py:1282]: This naming isn't perfect; the `model` here corresponds to Pyro's `model`, but I think the `guide` here corresponds to Pyro's `guide/model`.

File: funsor/montecarlo.py [hunk 2]
```diff
@@ -39,6 +41,25 @@ def monte_carlo_integrate(state, log_measure, integrand, reduced_vars):
...
+        sample_options["rng_key"], state.rng_key = jax.random.split(state.rng_key)
+
+    log_measure = model + guide
+    sample = log_measure.sample(approx_vars, state.sample_inputs, **sample_options)
+    if sample is log_measure:
+        return model  # cannot progress
+    reduced_vars = frozenset(
+        v for v in sample.input_vars if v.name in state.sample_inputs
+    )
+    result = (sample - guide).reduce(sum_op, reduced_vars)
```
Comment (fritzo) [inline funsor/montecarlo.py:59]: Here the Monte Carlo samples preserves the `model * guide`'s normalizing factor so we can simply divide by our `guide`; by contrast in Pyro variational samples are normalized to have unit norm so we must multiply by `model/guide`.

**Review by eb8680 — APPROVED** [review]
